File: python/ds/disjoint_set.py
# https://leetcode.com/explore/learn/card/graph/618/disjoint-set/3843/

import dataclasses


@dataclasses.dataclass
class DisjointSet:
    roots: dataclasses.field(default_factory=list)
    ranks: dataclasses.field(default_factory=list)

    def __init__(self, size):
        self.roots = [i for i in range(size)]
        self.ranks = [1] * size

    def find(self, x):
        if x == self.roots[x]:
            return x
        self.roots[x] = self.find(self.roots[x])
        return self.roots[x]

    def union(self, x, y):
        root_x, root_y = self.find(x), self.find(y)
        if root_x == root_y:
            return

        if self.ranks[root_x] > self.ranks[root_y]:
            self.roots[root_y] = root_x
        elif self.ranks[root_x] < self.ranks[root_y]:
            self.roots[root_x] = root_y
        else:
            self.roots[root_y] = root_x
            self.ranks[root_x] += 1

    def connected(self, x, y):
        return self.find(x) == self.find(y)


if __name__ == '__main__':
    _ds = DisjointSet(5)
    for _i in range(5):
        assert _i == _ds.find(_i)

    _ds.union(1, 3)
    _ds.union(2, 4)
    assert _ds.connected(1, 3)
    assert _ds.connected(2, 4)
    assert not _ds.connected(1, 2)
    assert not _ds.connected(0, 1)
    assert not _ds.connected(0, 2)

    _ds.union(0, 3)
    assert _ds.connected(1, 3)
    assert _ds.connected(2, 4)
    assert not _ds.connected(1, 2)
    assert _ds.connected(0, 1)
    assert not _ds.connected(0, 2)

    _ds.union(1, 2)
    assert _ds.connected(1, 3)
    assert _ds.connected(2, 4)
    assert _ds.connected(1, 2)
    assert _ds.connected(0, 1)
    assert _ds.connected(0, 2)

File: cpp/ds/disjointset.cpp
// https://leetcode.com/explore/learn/card/graph/618/disjoint-set/3843/
// http://e-maxx.ru/algo/dsu

#include <assert.h>
#include <vector>

using namespace std;

class DisjointSet {

private:
    int size;
    vector<int> roots;
    vector<int> ranks; // height of each vertex

public:
    DisjointSet(int _size): size(_size), roots(size), ranks(size) {
        for (int i = 0; i < size; ++i) {
            roots[i] = i;
            ranks[i] = 1;
        }
    }

    int find(int x) {
        if (x == roots[x]) return x;
        return roots[x] = find(roots[x]);
    }

    void unionn(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);

        if (rootX != rootY) {
            if (ranks[rootX] > ranks[rootY]) roots[rootY] = rootX;
            else if (ranks[rootX] < ranks[rootY]) roots[rootX] = rootY;
            else {
                roots[rootY] = rootX;
                ++ranks[rootX];
            }
            -- size;
        }
    }

    bool connected(int x, int y) {
        return find(x) == find(y);
    }

    int getSize() {
        return size;
    }
};


int main() {
    DisjointSet set(5);

    for (int i = 0; i < 5; ++i) assert(i == set.find(i));
    assert(5 == set.getSize());

    set.unionn(1, 3);
    set.unionn(2, 4);

    assert(set.connected(1, 3));
    assert(set.connected(2, 4));
    assert(!set.connected(1, 2));
    assert(!set.connected(0, 1));
    assert(!set.connected(0, 2));

    assert(3 == set.getSize());

    set.unionn(0, 3);

    assert(set.connected(1, 3));
    assert(set.connected(2, 4));
    assert(!set.connected(1, 2));
    assert(set.connected(0, 1));
    assert(!set.connected(0, 2));

    assert(2 == set.getSize());

    set.unionn(1, 2);

    assert(set.connected(0, 1));
    assert(set.connected(1, 2));
    assert(set.connected(2, 3));
    assert(set.connected(3, 4));
    assert(set.connected(4, 0));

    assert(1 == set.getSize());
}
